Rviz panel

heuristicus opened the pull request:
## Change Overview

This adds an rviz plugins package which contains a ros2 version of the spot panel I implemented for the spot_ros driver. The panel relies on the leases topic to enable and disable buttons so the user must specify a nonzero `lease_rate` in the config file.

![Screenshot from 2025-05-30 17-12-05](https://github.com/user-attachments/assets/2a3c51cc-becf-4c7b-9ed0-25c22158c86d)

Quite a bit of the functionality in the spot_ros driver is not available in the ros2 version, so I've disabled those buttons and commented out the code for the time being.

Currently service calls are done synchronously, which means the UI will be unresponsive until the service call completes. By default the call will timeout after 10 seconds, but this does not interrupt the call, so long-running requests like docking or undocking will still complete.

Any feedback is welcome.

## Testing Done

Started rviz and tested buttons for claim and release of lease, powering on and off, sitting and standing, docking and undocking, . Verified that the battery data, estimated runtime and battery temp is correctly displayed.

Untested:

- [x] Gentle estop
- [x] Kill motors
- [x] Stop
- [x] Release e-stop (for software stops)
- [x] Self-right

Todo:

- [x] Make service calls asynchronous. Need to work out how to enable/disable the release/claim buttons depending on the result of the async call.

heuristicus commented:
I can definitely remove the commented stuff and add in later if needed. I think there's some stuff which is no longer exposed so I'd have to go through and work out what is feasible to add and what isn't. The cam components I'd probably remove since I don't have access to it anymore.

I think the most useful components rely on the services below, but in my brief pass I couldn't see them any more. As far as I know they're still available through the wrapper but it's a question of whether people would find it beneficial to have access to these. In my experience most of the components which I put into tabs are much less used, except for perhaps docking and velocity settings.

```
  gaitService_ = nh_.serviceClient<spot_msgs::SetLocomotion>("/spot/locomotion_mode");
  swingHeightService_ = nh_.serviceClient<spot_msgs::SetSwingHeight>("/spot/swing_height");
  terrainParamsService_ = nh_.serviceClient<spot_msgs::SetTerrainParams>("/spot/terrain_params");
  obstacleParamsService_ = nh_.serviceClient<spot_msgs::SetObstacleParams>("/spot/obstacle_params");
```

tcappellari-bdai commented on spot_rviz_plugins/src/spot_panel.cpp:
```diff
@@ -0,0 +1,812 @@
+  haveLease = false;
+  motionAllowed = false;
+
+  sitService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/sit");
+  standService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/stand");
+  claimLeaseService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/claim");
+  releaseLeaseService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/release");
+  powerOnService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/power_on");
+  powerOffService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/power_off");
+  maxVelocityService_ = client_node_->create_client<spot_msgs::srv::SetVelocity>("/spot/velocity_limit");
+  hardStopService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/estop/hard");
+  gentleStopService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/estop/gentle");
+  releaseStopService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/estop/release");
+  stopService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/stop");
+  // gaitService_ = nh_.serviceClient<spot_msgs::SetLocomotion>("/spot/locomotion_mode");
+  // swingHeightService_ = nh_.serviceClient<spot_msgs::SetSwingHeight>("/spot/swing_height");
+  // terrainParamsService_ = nh_.serviceClient<spot_msgs::SetTerrainParams>("/spot/terrain_params");
+  // obstacleParamsService_ = nh_.serviceClient<spot_msgs::SetObstacleParams>("/spot/obstacle_params");
+  // allowMotionService_ = nh_.serviceClient<std_srvs::SetBool>("/spot/allow_motion");
+  // bodyPoseService_ = nh_.serviceClient<spot_msgs::PosedStand>("/spot/posed_stand");
+  dockService_ = client_node_->create_client<spot_msgs::srv::Dock>("/spot/dock");
+  undockService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/undock");
+  selfRightService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/self_right");
+  // rollOverRightService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/roll_over_right");
+  // rollOverLeftService_ = client_node_->create_client<std_srvs::srv::Trigger>("/spot/roll_over_left");
+
+  // camSetScreenService_ = nh_.serviceClient<spot_cam::SetString>("/spot/cam/set_screen");
+  // camSetPTZService_ = nh_.serviceClient<spot_cam::SetPTZState>("/spot/cam/ptz/set_position");
+  // camLookAtPointService_ = nh_.serviceClient<spot_cam::LookAtPoint>("/spot/cam/ptz/look_at_point");
```
I was able to try out your panel and got it up on my rviz but realized all the topic names here are hardcoded with `/spot/` as the namespace, which isn't compatible with our workflow. We have a parameter in the driver `spot_name` that stores the name of the robot we are connected to that we use to populate the namespaces of all relevant topics and services. Could you make a change to use the parameterized spot name instead of the hardcoded `spot`?

Besides that, I would just also ask to remove any unnecessary dead code

heuristicus commented on spot_rviz_plugins/src/spot_panel.cpp:
```diff
@@ -0,0 +1,812 @@
+    std::stringstream stream;
+    stream << std::fixed << std::setprecision(0) << battState.charge_percentage;
+    std::string pct = stream.str() + "%";
+    stream.str("");
```
From what I can tell, because I'm reusing the stream afterwards I need to call `str("")` to clean out the old data. `clear` only resets the error flags on the stream.

heuristicus commented:
I've removed unused code, updated the trigger request function template, and added a line edit which allows you to specify the robot's name. I think this is the best I can do in terms of connecting to the robot with the correct name. Unfortunately the content of the line edit isn't saved if you save the rviz config so it has to be set every time. If someone knows how to make a setting stick that would be useful to know.

heuristicus commented:
Thanks for the tests, Katie! All button states are currently enabled or disabled based on whether the Ros driver has the lease or not. If the driver doesn't have the lease you can press the claim button to claim it. Do you interact with the robot in a way that would work better with different behaviour?

khughes-bdai commented:
@heuristicus Hmm, I was hitting the claim button and this was succeeding, but wasn't getting the power on buttons back 

[Screencast from 07-08-2025 02:19:02 PM.webm](https://github.com/user-attachments/assets/af123a22-e201-439c-9b42-12c72548da3d)

heuristicus commented:
I've just taken a look at this and managed to reproduce the issue. I think it might have been something to do with how I'm checking whether the lease is owned by ros. I assumed that `ros_spot` (the `SPOT_CLIENT_NAME` defined in spot_wrapper/wrapper.py) was at the beginning of the lease owner string, but perhaps this was not the case for some reason. I've modified it so that if `ros_spot` appears anywhere in the string the lease is assumed to be owned by ROS and thus things should be enabled.

khughes-bdai commented:
@heuristicus Thanks for looking into this! Will test on robot again and let you know!

heuristicus commented:
I'm using an older version of everything to ensure compatibility with some robots we are using which are still on older versions of the spot SDK, specifically 5c05a576e19983c284b6db4de628c69b67361556. With that one there are continual messages on the lease topic even if nobody is holding the lease. Could the issue be related to https://github.com/bdaiinstitute/spot_wrapper/pull/174 that just got merged?

heuristicus commented:
@tcappellari-bdai what I meant is that maybe @khughes-bdai is using a version of the wrapper which isn't using that PR, in which case the problem with the lease messages not coming through might be a result of that code being missing. I think because I'm using earlier code the lease publishing works fine for me.

khughes-bdai commented:
Okay yes that was exactly my issue -- when I correctly updated to that spot wrapper commit via https://github.com/bdaiinstitute/spot_ros2/pull/721 it is working great!!

khughes-bdai reviewed (approved):
LGTM! Thanks @heuristicus

heuristicus commented:
Thanks for your help getting this merged in @khughes-bdai! As a final addition I updated the default rviz config and template to include the panel. If that's not OK I can revert.
